subxt/src/storage/storage_client.rs

```diff
@@ -78,339 +64,29 @@ where
-
-            Ok(KeyIter {
-                client,
-                address_root_bytes,
-                metadata,
-                return_type_id,
-                block_hash: hash,
-                count: page_size,
-                start_key: None,
-                buffer: Default::default(),
-                _marker: std::marker::PhantomData,
-            })
-        }
-    }
-}
-
-/// Iterates over key value pairs in a map.
-pub struct KeyIter<T: Config, Client, ReturnTy> {
-    client: StorageClient<T, Client>,
-    address_root_bytes: Vec<u8>,
-    return_type_id: u32,
-    metadata: Metadata,
-    count: u32,
-    block_hash: T::Hash,
-    start_key: Option<StorageKey>,
-    buffer: Vec<(StorageKey, StorageData)>,
-    _marker: std::marker::PhantomData<ReturnTy>,
-}
-
-impl<'a, T, Client, ReturnTy> KeyIter<T, Client, ReturnTy>
-where
-    T: Config,
-    Client: OnlineClientT<T>,
-    ReturnTy: DecodeWithMetadata,
-{
-    /// Returns the next key value pair from a map.
-    pub async fn next(
-        &mut self,
-    ) -> Result<Option<(StorageKey, ReturnTy::Target)>, Error> {
-        loop {
-            if let Some((k, v)) = self.buffer.pop() {
-                let val = ReturnTy::decode_with_metadata(
-                    &mut &v.0[..],
-                    self.return_type_id,
-                    &self.metadata,
-                )?;
-                return Ok(Some((k, val)))
-            } else {
-                let start_key = self.start_key.take();
-                let keys = self
-                    .client
-                    .fetch_keys(
-                        &self.address_root_bytes,
-                        self.count,
-                        start_key.as_ref().map(|k| &*k.0),
-                        Some(self.block_hash),
-                    )
-                    .await?;
-
-                if keys.is_empty() {
-                    return Ok(None)
-                }
-
-                self.start_key = keys.last().cloned();
-
-                let change_sets = self
-                    .client
-                    .client
-                    .rpc()
-                    .query_storage_at(keys.iter().map(|k| &*k.0), Some(self.block_hash))
-                    .await?;
-                for change_set in change_sets {
-                    for (k, v) in change_set.changes {
-                        if let Some(v) = v {
-                            self.buffer.push((k, v));
-                        }
-                    }
+            // If block hash is not provided, get the hash
+            // for the latest block and use that.
+            let block_hash = match block_hash {
+                Some(hash) => hash,
+                None => {
+                    client
+                        .rpc()
+                        .block_hash(None)
+                        .await?
+                        .expect("didn't pass a block number; qed")
```

Comment:
Do we need this?

What I mean is; all of the "old" storage methods took an `Option<T::Hash>` as an arg and didn't do this; they just passed it into the relevant RPC method as-is. And if we didn't need to do this, we'd both save an API call and be able to remove the `.await?` from all of the `api.storage().at()` calls :)

There might need to be something like this for the new API, but we could have a little helper function which resolves the `Option<T::Hash>` as part of each storage call, so that the `.await?` there can handle it all :)

Reply:
But that all said, I can see that it's nice how aligned this is with how `Events` work....